pipeline handler expecting dest_path for file to archive

testing the ANFOG_DM pipeline with a ZIP containing a netcdf file a jpg to upload to S3 and a zip file to archive.
The handler fails processing the file to archive. It is expecting a non-empty string for the dest-path 

 `Testing started at 12:19 PM ...
/Users/bpasquer/Documents/CHEF/chef/src/python-aodndata/python-aodndata-virtualenv/bin/python "/Applications/PyCharm CE.app/Contents/helpers/pydev/pydevd.py" --multiproc --qt-support=auto --client 127.0.0.1 --port 58763 --file "/Applications/PyCharm CE.app/Contents/helpers/pycharm/_jb_unittest_runner.py" --target test_anfog.TestAnfogHandler.test_good_anfog_dm_zip
pydev debugger: process 9868 is connecting

Connected to pydev debugger (build 173.4301.16)
Launching unittests with arguments python -m unittest test_anfog.TestAnfogHandler.test_good_anfog_dm_zip in /Users/bpasquer/Documents/CHEF/chef/src/python-aodndata/test_aodndata/anfog
2018-03-06 12:19:58,812 WARNING [root] no logger parameter or celery task found, falling back to root logger
2018-03-06 12:19:58,816 INFO [root] running handler -> 'AnfogHandler({'collection_dir': '/var/folders/1q/r320d1956jq3gqw5rtfd6cg06z0c81/T/TestAnfogHandler0kJ9L8/AnfogHandlerAxXXxe/collection', 'dest_path_function': None, 'file_type': <FileType.ZIP: ('.zip', 'application/zip')>, 'should_notify': None, 'resolve_params': None, 'file_checksum': None, 'instance_working_directory': '/var/folders/1q/r320d1956jq3gqw5rtfd6cg06z0c81/T/TestAnfogHandler0kJ9L8/AnfogHandlerAxXXxe', 'celery_task_id': None, 'result': <HandlerResult.UNKNOWN: 0>, 'exclude_regexes': None, 'include_regexes': None, 'cc_versions': None, 'products_dir': '/var/folders/1q/r320d1956jq3gqw5rtfd6cg06z0c81/T/TestAnfogHandler0kJ9L8/AnfogHandlerAxXXxe/products', 'check_params': {'checks': ['cf']}, 'notification_results': None, 'error_details': None, 'temp_dir': '/var/folders/1q/r320d1956jq3gqw5rtfd6cg06z0c81/T/TestAnfogHandler0kJ9L8/AnfogHandlerAxXXxe/temp', 'file_extension': '.zip', 'archive_path_function': None, 'archive_input_file': False, 'input_file': '/Users/bpasquer/Documents/CHEF/chef/src/python-aodndata/test_aodndata/anfog/BassStrait20160302.zip', 'start_time': datetime.datetime(2018, 3, 6, 12, 19, 58, 809307), 'harvest_params': None, 'upload_path': None, 'harvest_type': 'talend', 'celery_task_name': 'NO_TASK', 'notify_params': None, 'file_collection': PipelineFileCollection([]), 'allowed_extensions': ['.nc', '.kml', '.pdf', '.txt', '.png', '.jpeg', '.zip'], 'is_archived': False, 'error': None})'
2018-03-06 12:19:59,256 SYSINFO [root] get_file_checksum -> 'cbc824207d25672ca3ed5fd482f980109c06286a37a23e60a523f558476cf5eb'
2018-03-06 12:19:59,257 SYSINFO [root] get_cc_module_versions -> {'core': '2.3.1', 'plugin': '1.1.3'}
2018-03-06 12:19:59,258 SYSINFO [root] get_path_function -> 'AnfogHandler.dest_path'
2018-03-06 12:19:59,259 SYSINFO [root] get_path_function (archive) -> 'AnfogHandler.dest_path'
2018-03-06 12:19:59,260 SYSINFO [root] AnfogHandler transitioned to state: HANDLER_INITIALISED
2018-03-06 12:19:59,261 SYSINFO [root] get_resolve_runner -> 'ZipFileResolveRunner'
2018-03-06 12:20:00,602 INFO [root] file: 'BassStrait20160302_PSALjpeg.jpg' updated: {'publish_type': 'HARVEST_UPLOAD'}
2018-03-06 12:20:00,603 INFO [root] file: 'BassStrait20160302_rawfiles.zip' updated: {'publish_type': 'HARVEST_UPLOAD'}
2018-03-06 12:20:00,604 INFO [root] file: 'IMOS_ANFOG_BCEOPSTUV_20160301T234750Z_SL286_FV01_timeseries_END-20160321T002220Z.nc' updated: {'publish_type': 'HARVEST_UPLOAD'}
2018-03-06 12:20:00,605 SYSINFO [root] AnfogHandler transitioned to state: HANDLER_RESOLVED
2018-03-06 12:20:00,642 INFO [root] file: 'IMOS_ANFOG_BCEOPSTUV_20160301T234750Z_SL286_FV01_timeseries_END-20160321T002220Z.nc' updated: {'dest_path': 'IMOS/ANFOG/slocum_glider/BassStrait20160302/IMOS_ANFOG_BCEOPSTUV_20160301T234750Z_SL286_FV01_timeseries_END-20160321T002220Z.nc'}
2018-03-06 12:20:00,644 INFO [root] file: 'IMOS_ANFOG_BCEOPSTUV_20160301T234750Z_SL286_FV01_timeseries_END-20160321T002220Z.nc' updated: {'check_type': 'NC_COMPLIANCE_CHECK'}
2018-03-06 12:20:00,650 INFO [root] file: 'BassStrait20160302_rawfiles.zip' updated: {'check_type': 'FORMAT_CHECK'}
2018-03-06 12:20:00,652 INFO [root] file: 'BassStrait20160302_rawfiles.zip' updated: {'publish_type': 'ARCHIVE_ONLY'}
2018-03-06 12:20:00,656 INFO [root] file: 'BassStrait20160302_rawfiles.zip' updated: {'archive_path': 'IMOS/ANFOG/slocum_glider/BassStrait20160302/BassStrait20160302_rawfiles.zip'}
2018-03-06 12:20:00,659 INFO [root] file: 'BassStrait20160302_PSALjpeg.jpg' updated: {'check_type': 'FORMAT_CHECK'}
2018-03-06 12:20:00,660 INFO [root] file: 'BassStrait20160302_PSALjpeg.jpg' updated: {'publish_type': 'UPLOAD_ONLY'}
2018-03-06 12:20:00,661 INFO [root] file: 'BassStrait20160302_PSALjpeg.jpg' updated: {'dest_path': 'IMOS/ANFOG/slocum_glider/BassStrait20160302/BassStrait20160302_PSALjpeg.jpg'}
2018-03-06 12:20:00,662 SYSINFO [root] AnfogHandler transitioned to state: HANDLER_PREPROCESSED
2018-03-06 12:20:00,663 SYSINFO [root] get_check_runner -> 'CheckRunnerAdapter'
2018-03-06 12:20:00,664 INFO [root] file: 'BassStrait20160302_PSALjpeg.jpg' updated: {'check_type': 'FORMAT_CHECK'}
2018-03-06 12:20:00,665 INFO [root] file: 'BassStrait20160302_rawfiles.zip' updated: {'check_type': 'FORMAT_CHECK'}
2018-03-06 12:20:00,666 INFO [root] file: 'IMOS_ANFOG_BCEOPSTUV_20160301T234750Z_SL286_FV01_timeseries_END-20160321T002220Z.nc' updated: {'check_type': 'NC_COMPLIANCE_CHECK'}
2018-03-06 12:20:00,667 SYSINFO [root] get_child_check_runner -> 'FormatCheckRunner'
2018-03-06 12:20:00,668 INFO [root] checking that '/var/folders/1q/r320d1956jq3gqw5rtfd6cg06z0c81/T/TestAnfogHandler0kJ9L8/AnfogHandlerAxXXxe/collection/BassStrait20160302_rawfiles.zip' is not empty
2018-03-06 12:20:00,669 INFO [root] file: 'BassStrait20160302_rawfiles.zip' updated: {'is_checked': True}
2018-03-06 12:20:00,671 INFO [root] checking that '/var/folders/1q/r320d1956jq3gqw5rtfd6cg06z0c81/T/TestAnfogHandler0kJ9L8/AnfogHandlerAxXXxe/collection/BassStrait20160302_PSALjpeg.jpg' is not empty
2018-03-06 12:20:00,672 INFO [root] file: 'BassStrait20160302_PSALjpeg.jpg' updated: {'is_checked': True}
Could not load imos13 = cc_plugin_imos.imos:IMOS1_3Check : (netCDF4 1.3.1 (/Users/bpasquer/Documents/CHEF/chef/src/python-aodndata/python-aodndata-virtualenv/lib/python2.7/site-packages), Requirement.parse('netCDF4==1.2.4'))
Could not load imos14 = cc_plugin_imos.imos:IMOS1_4Check : (netCDF4 1.3.1 (/Users/bpasquer/Documents/CHEF/chef/src/python-aodndata/python-aodndata-virtualenv/lib/python2.7/site-packages), Requirement.parse('netCDF4==1.2.4'))
Could not load ghrsst = cc_plugin_imos.srs:IMOSGHRSSTCheck : (netCDF4 1.3.1 (/Users/bpasquer/Documents/CHEF/chef/src/python-aodndata/python-aodndata-virtualenv/lib/python2.7/site-packages), Requirement.parse('netCDF4==1.2.4'))
2018-03-06 12:20:00,893 SYSINFO [root] get_child_check_runner -> 'ComplianceCheckerCheckRunner'
2018-03-06 12:20:00,893 INFO [root] checking compliance of '/var/folders/1q/r320d1956jq3gqw5rtfd6cg06z0c81/T/TestAnfogHandler0kJ9L8/AnfogHandlerAxXXxe/collection/IMOS_ANFOG_BCEOPSTUV_20160301T234750Z_SL286_FV01_timeseries_END-20160321T002220Z.nc' against ['cf']
2018-03-06 12:20:01,186 INFO [root] file: 'IMOS_ANFOG_BCEOPSTUV_20160301T234750Z_SL286_FV01_timeseries_END-20160321T002220Z.nc' updated: {'is_checked': True}
2018-03-06 12:20:01,188 SYSINFO [root] AnfogHandler transitioned to state: HANDLER_CHECKED
2018-03-06 12:20:01,189 SYSINFO [root] `process` not overridden by child class, skipping step
2018-03-06 12:20:01,191 SYSINFO [root] AnfogHandler transitioned to state: HANDLER_PROCESSED
2018-03-06 12:20:01,193 SYSINFO [root] get_upload_runner (archive) -> 'FileUploadRunner'
2018-03-06 12:20:01,194 INFO [root] uploading '/var/folders/1q/r320d1956jq3gqw5rtfd6cg06z0c81/T/TestAnfogHandler0kJ9L8/AnfogHandlerAxXXxe/collection/BassStrait20160302_rawfiles.zip' to 'file:///tmp/probably/doesnt/exist/archive/IMOS/ANFOG/slocum_glider/BassStrait20160302/BassStrait20160302_rawfiles.zip'
2018-03-06 12:20:01,197 INFO [root] file: 'BassStrait20160302_rawfiles.zip' updated: {'is_archived': True}
2018-03-06 12:20:01,197 SYSINFO [root] get_upload_runner -> 'FileUploadRunner'
2018-03-06 12:20:29,120 INFO [root] file: 'BassStrait20160302_PSALjpeg.jpg' updated: {'is_overwrite': False}
2018-03-06 12:22:47,955 ERROR [root] KeyboardInterrupt: 
Traceback (most recent call last):
  File "/Users/bpasquer/Documents/CHEF/chef/src/python-aodndata/python-aodndata-virtualenv/lib/python2.7/site-packages/aodncore/pipeline/handlerbase.py", line 777, in run
    self.trigger(transition['trigger'])
  File "/Users/bpasquer/Documents/CHEF/chef/src/python-aodndata/python-aodndata-virtualenv/lib/python2.7/site-packages/transitions/core.py", line 33, in get_trigger
    return func(*args, **kwargs)
  File "/Users/bpasquer/Documents/CHEF/chef/src/python-aodndata/python-aodndata-virtualenv/lib/python2.7/site-packages/transitions/core.py", line 283, in trigger
    return self.machine._process(f)
  File "/Users/bpasquer/Documents/CHEF/chef/src/python-aodndata/python-aodndata-virtualenv/lib/python2.7/site-packages/transitions/core.py", line 840, in _process
    return trigger()
  File "/Users/bpasquer/Documents/CHEF/chef/src/python-aodndata/python-aodndata-virtualenv/lib/python2.7/site-packages/transitions/core.py", line 316, in _trigger
    except Exception as e:
  File "/Users/bpasquer/Documents/CHEF/chef/src/python-aodndata/python-aodndata-virtualenv/lib/python2.7/site-packages/transitions/core.py", line 316, in _trigger
    except Exception as e:
  File "_pydevd_bundle/pydevd_cython_darwin_27_64.pyx", line 968, in _pydevd_bundle.pydevd_cython_darwin_27_64.SafeCallWrapper.__call__
  File "_pydevd_bundle/pydevd_cython_darwin_27_64.pyx", line 515, in _pydevd_bundle.pydevd_cython_darwin_27_64.PyDBFrame.trace_dispatch
  File "_pydevd_bundle/pydevd_cython_darwin_27_64.pyx", line 880, in _pydevd_bundle.pydevd_cython_darwin_27_64.PyDBFrame.trace_dispatch
  File "_pydevd_bundle/pydevd_cython_darwin_27_64.pyx", line 843, in _pydevd_bundle.pydevd_cython_darwin_27_64.PyDBFrame.trace_dispatch
  File "_pydevd_bundle/pydevd_cython_darwin_27_64.pyx", line 254, in _pydevd_bundle.pydevd_cython_darwin_27_64.PyDBFrame.do_wait_suspend
  File "/Applications/PyCharm CE.app/Contents/helpers/pydev/pydevd.py", line 817, in do_wait_suspend
    time.sleep(0.01)
KeyboardInterrupt
2018-03-06 12:22:47,961 SYSINFO [root] get_notify_runner -> 'NotifyRunnerAdapter'
2018-03-06 12:22:47,963 SYSINFO [root] AnfogHandler transitioned to state: HANDLER_NOTIFIED_ERROR
2018-03-06 12:22:47,964 INFO [root] handler result for input_file '/Users/bpasquer/Documents/CHEF/chef/src/python-aodndata/test_aodndata/anfog/BassStrait20160302.zip': ERROR
2018-03-06 12:22:47,965 SYSINFO [root] AnfogHandler transitioned to state: HANDLER_COMPLETED_ERROR


Failure
Traceback (most recent call last):
  File "/System/Library/Frameworks/Python.framework/Versions/2.7/lib/python2.7/unittest/case.py", line 331, in run
    testMethod()
  File "/Users/bpasquer/Documents/CHEF/chef/src/python-aodndata/test_aodndata/anfog/test_anfog.py", line 42, in test_good_anfog_dm_zip
    handler = self.run_handler(GOOD_ZIP_DM, check_params={'checks': ['cf']})
  File "/Users/bpasquer/Documents/CHEF/chef/src/python-aodndata/python-aodndata-virtualenv/lib/python2.7/site-packages/aodncore/testlib/handlertest.py", line 38, in run_handler
    handler.run()
  File "/System/Library/Frameworks/Python.framework/Versions/2.7/lib/python2.7/unittest/case.py", line 956, in assertIsNone
    self.fail(self._formatMessage(msg, standardMsg))
  File "/System/Library/Frameworks/Python.framework/Versions/2.7/lib/python2.7/unittest/case.py", line 412, in fail
    raise self.failureException(msg)
AssertionError: KeyboardInterrupt() is not None


Ran 1 test in 169.197s

FAILED (failures=1)

Process finished with exit code 1`

--- a/aodncore/pipeline/exceptions.py
+++ b/aodncore/pipeline/exceptions.py
@@ -4,6 +4,7 @@ __all__ = [
     'PipelineProcessingError',
     'PipelineSystemError',
     'ComplianceCheckFailedError',
+    'AttributeNotSetError',
     'DuplicatePipelineFileError',
     'FileDeleteFailedError',
     'FileUploadFailedError',
@@ -61,6 +62,10 @@ class InvalidFileFormatError(PipelineProcessingError):
 
 
 # System errors
+
+class AttributeNotSetError(PipelineSystemError):
+    pass
+
 
 class DuplicatePipelineFileError(PipelineSystemError):
     pass

--- a/aodncore/pipeline/steps/upload.py
+++ b/aodncore/pipeline/steps/upload.py
@@ -7,7 +7,7 @@ from paramiko import SSHClient, AutoAddPolicy
 from six.moves.urllib.parse import urlparse
 
 from .basestep import AbstractCollectionStepRunner
-from ..exceptions import FileDeleteFailedError, FileUploadFailedError, InvalidUploadUrlError
+from ..exceptions import AttributeNotSetError, FileDeleteFailedError, FileUploadFailedError, InvalidUploadUrlError
 from ..files import validate_pipelinefilecollection
 from ...util import format_exception, mkdir_p, rm_f, safe_copy_file
 
@@ -89,12 +89,23 @@ class BaseUploadRunner(AbstractCollectionStepRunner):
         pass
 
     @abc.abstractmethod  # pragma: no cover
-    def set_is_overwrite(self, pipeline_files):
+    def _get_is_overwrite(self, pipeline_file, abs_path):
         pass
 
     def _get_absolute_dest_path(self, pipeline_file):
         rel_path = getattr(pipeline_file, self.dest_path_attr)
+        if not rel_path:
+            raise AttributeNotSetError(
+                "attribute '{attr}' not set in '{pf}'".format(attr=self.dest_path_attr, pf=pipeline_file))
         return os.path.join(self.prefix, rel_path)
+
+    def set_is_overwrite(self, pipeline_files):
+        validate_pipelinefilecollection(pipeline_files)
+
+        should_upload = pipeline_files.filter_by_bool_attributes_and_not('should_store', 'is_deletion')
+        for pipeline_file in should_upload:
+            abs_path = self._get_absolute_dest_path(pipeline_file)
+            pipeline_file.is_overwrite = self._get_is_overwrite(pipeline_file, abs_path)
 
     def run(self, pipeline_files):
         validate_pipelinefilecollection(pipeline_files)
@@ -156,6 +167,9 @@ class FileUploadRunner(BaseUploadRunner):
     def _get_absolute_dest_uri(self, pipeline_file):
         return "file://{path}".format(path=self._get_absolute_dest_path(pipeline_file))
 
+    def _get_is_overwrite(self, pipeline_file, abs_path):
+        return os.path.exists(abs_path)
+
     def _post_run_hook(self):
         return
 
@@ -166,13 +180,6 @@ class FileUploadRunner(BaseUploadRunner):
         abs_path = self._get_absolute_dest_path(pipeline_file)
         mkdir_p(os.path.dirname(abs_path))
         safe_copy_file(pipeline_file.src_path, abs_path, overwrite=True)
-
-    def set_is_overwrite(self, pipeline_files):
-        validate_pipelinefilecollection(pipeline_files)
-
-        for pipeline_file in pipeline_files:
-            abs_path = self._get_absolute_dest_path(pipeline_file)
-            pipeline_file.is_overwrite = os.path.exists(abs_path)
 
 
 class S3UploadRunner(BaseUploadRunner):
@@ -202,6 +209,10 @@ class S3UploadRunner(BaseUploadRunner):
     def _get_absolute_dest_uri(self, pipeline_file):
         return "s3://{bucket}/{path}".format(bucket=self.bucket, path=self._get_absolute_dest_path(pipeline_file))
 
+    def _get_is_overwrite(self, pipeline_file, abs_path):
+        response = self.s3_client.list_objects_v2(Bucket=self.bucket, Prefix=abs_path)
+        return bool(k for k in response.get('Contents', []) if k['Key'] == abs_path)
+
     def _post_run_hook(self):
         return
 
@@ -221,14 +232,6 @@ class S3UploadRunner(BaseUploadRunner):
         except Exception as e:
             raise InvalidUploadUrlError(
                 "unable to access S3 bucket '{0}': {1}".format(self.bucket, format_exception(e)))
-
-    def set_is_overwrite(self, pipeline_files):
-        validate_pipelinefilecollection(pipeline_files)
-
-        for pipeline_file in pipeline_files:
-            abs_path = self._get_absolute_dest_path(pipeline_file)
-            response = self.s3_client.list_objects_v2(Bucket=self.bucket, Prefix=abs_path)
-            pipeline_file.is_overwrite = bool(k for k in response.get('Contents', []) if k['Key'] == abs_path)
 
 
 def sftp_path_exists(sftpclient, path):
@@ -319,6 +322,9 @@ class SftpUploadRunner(BaseUploadRunner):
     def _get_absolute_dest_uri(self, pipeline_file):
         return "sftp://{server}{path}".format(server=self.server, path=self._get_absolute_dest_path(pipeline_file))
 
+    def _get_is_overwrite(self, pipeline_file, abs_path):
+        return sftp_path_exists(self.sftp_client, abs_path)
+
     def _post_run_hook(self):
         return
 
@@ -332,10 +338,3 @@ class SftpUploadRunner(BaseUploadRunner):
 
         with open(pipeline_file.src_path, 'rb') as f:
             self.sftp_client.putfo(f, abs_path, confirm=True)
-
-    def set_is_overwrite(self, pipeline_files):
-        validate_pipelinefilecollection(pipeline_files)
-
-        for pipeline_file in pipeline_files:
-            abs_path = self._get_absolute_dest_path(pipeline_file)
-            pipeline_file.is_overwrite = sftp_path_exists(self.sftp_client, abs_path)

--- a/aodncore/testlib/testutil.py
+++ b/aodncore/testlib/testutil.py
@@ -58,8 +58,8 @@ class NullUploadRunner(BaseUploadRunner):
     def _get_absolute_dest_uri(self, pipeline_file):
         return "null://{dest_path}".format(dest_path=pipeline_file.dest_path)
 
-    def set_is_overwrite(self, pipeline_files):
-        pass
+    def _get_is_overwrite(self, pipeline_file, abs_path):
+        return not self.fail
 
     def run(self, pipeline_files):
         self.call_count += 1
@@ -74,8 +74,7 @@ class NullUploadRunner(BaseUploadRunner):
             raise AssertionError("run method not called")
 
     def assert_not_called(self):
-        if self.call_count != 0:
-            raise AssertionError("run method call_count: {call_count}".format(call_count=self.call_count))
+        self.assert_call_count(0)
 
 
 def dest_path_testing(filename):

--- a/test_aodncore/pipeline/steps/test_upload.py
+++ b/test_aodncore/pipeline/steps/test_upload.py
@@ -205,6 +205,21 @@ class TestBaseUploadRunner(BaseTestCase):
         runner.run(collection)
         self.assertTrue(collection[0].is_stored)
 
+    def test_set_is_overwrite_with_no_action_file(self):
+        collection = get_upload_collection()
+
+        temp_file = PipelineFile(self.temp_nc_file)
+        self.assertIsNone(temp_file.dest_path)
+        self.assertIs(temp_file.publish_type, PipelineFilePublishType.NO_ACTION)
+        collection.add(temp_file)
+
+        runner = NullUploadRunner("/")
+        try:
+            runner.set_is_overwrite(collection)
+        except Exception as e:
+            raise AssertionError(
+                "unexpected exception raised. {cls} {msg}".format(cls=e.__class__.__name__, msg=e))
+
 
 class TestFileUploadRunner(BaseTestCase):
     @mock.patch('aodncore.pipeline.steps.upload.mkdir_p')

